src/field_theory/fixed.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import algebra.group_action_hom
+import field_theory.normal
+import field_theory.subfield
+import ring_theory.polynomial
+
+/-!
+# Fixed field under a group action.
+
+This is the basis of the Fundamental Theorem of Galois Theory.
+Given a (finite) group `G` that acts on a field `F`, we define `fixed_points G F`,
+the subfield consisting of elements of `F` fixed_points by every element of `G`.
+
+This subfield is then normal (proved in this file) and separable (TODO), and in addition
+if `G` acts faithfully on `F` then `findim (fixed_points G F) F = fintype.card G` (TODO).
+
+## Main Deifnitions
+
+- `fixed_points G F`, the subfield consisting of elements of `F` fixed_points by every element of `G`, where
+`G` is a group that acts on `F`.
+
+-/
+
+noncomputable theory
+open_locale classical
+open mul_action
+
+universes u v w
+
+variables (G : Type u) [group G] (F : Type v) [field F] [mul_semiring_action G F] (g : G)
+
+/-- The subfield fixed_points by one element of the group. -/
+def fixed_by : set F :=
+{ x | g • x = x }
+
+instance fixed_by.is_subfield : is_subfield (fixed_by G F g) :=
+{ zero_mem := smul_zero g,
+  add_mem := λ x y hx hy, (smul_add g x y).trans $ congr_arg2 _ hx hy,
+  neg_mem := λ x hx, (smul_neg g x).trans $ congr_arg _ hx,
+  one_mem := smul_one g,
+  mul_mem := λ x y hx hy, (smul_mul' g x y).trans $ congr_arg2 _ hx hy,
+  inv_mem := λ x hx, (smul_inv F g x).trans $ congr_arg _ hx }
+
+theorem fixed_eq_Inter_fixed_by : fixed_points G F = ⋂ g : G, fixed_by G F g :=
+set.ext $ λ x, ⟨λ hx, set.mem_Inter.2 $ λ g, hx g, λ hx g, by convert set.mem_Inter.1 hx g; refl⟩
+
+instance fixed_points.is_subfield : is_subfield (fixed_points G F) :=
+by convert @is_subfield.Inter F _ G (fixed_by G F) _; rw fixed_eq_Inter_fixed_by
+
+instance fixed_points.is_invariant_subring : is_invariant_subring G (fixed_points G F) :=
+{ smul_mem := λ g x hx g', by rw [hx, hx] }
+
+theorem fixed_points.smul (g : G) (x : fixed_points G F) : g • x = x :=
+subtype.eq $ x.2 g
+
+theorem fixed_points.smul_polynomial (g : G) (p : polynomial (fixed_points G F)) : g • p = p :=
+polynomial.induction_on p
+  (λ x, by rw [polynomial.smul_C, fixed_points.smul])
+  (λ p q ihp ihq, by rw [smul_add, ihp, ihq])
+  (λ n x ih, by rw [smul_mul', polynomial.smul_C, fixed_points.smul, smul_pow, polynomial.smul_X])
+
+instance fixed_points.algebra : algebra (fixed_points G F) F :=
+algebra.of_subring _
+
+theorem fixed_points.coe_algebra_map :
+  (algebra_map (fixed_points G F) F) = is_subring.subtype (fixed_points G F) :=
+rfl
+
+variables [fintype G] (x : F)
+
+def fixed_points.polynomial : polynomial (fixed_points G F) :=
```

Comment:
Should this thing be called `minpoly`, maybe?

Reply:
done